src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -121,12 +142,70 @@ absl::flat_hash_set<NodeID> GcsResourceScheduler::FilterCandidateNodes(
   return result;
 }
 
-const std::vector<ResourceSet> &GcsResourceScheduler::SortRequiredResources(
+std::tuple<bool, bool> cmp_resource(
+    const std::string &resource,
+    const absl::flat_hash_map<std::string, FixedPoint> &a_map,
+    const absl::flat_hash_map<std::string, FixedPoint> &b_map) {
+  auto a_resource = a_map.find(resource);
+  auto b_resource = b_map.find(resource);
+
+  bool a_less_than_b = false;
+  bool finished = false;
+
+  if (a_resource != a_map.end() && b_resource != b_map.end()) {
+    if (a_resource->second != b_resource->second) {
+      a_less_than_b = a_resource->second < b_resource->second;
+      finished = true;
+    }
+  } else if (a_resource != a_map.end() && a_resource->second != 0) {
+    a_less_than_b = false;
+    finished = true;
+  } else if (b_resource != b_map.end() && b_resource->second != 0) {
+    a_less_than_b = true;
+    finished = true;
+  }
+  return std::make_tuple(finished, a_less_than_b);
+}
+
+std::vector<size_t> GcsResourceScheduler::SortRequiredResources(
     const std::vector<ResourceSet> &required_resources) {
-  // TODO(ffbin): A bundle may require special resources, such as GPU. We need to
-  // schedule bundles with special resource requirements first, which will be implemented
-  // in the next pr.
-  return required_resources;
+  std::vector<size_t> sorted_index(required_resources.size());
+  std::iota(sorted_index.begin(), sorted_index.end(), 0);
+  std::sort(sorted_index.begin(), sorted_index.end(), [required_resources](int b, int a) {
```

Comment:
```suggestion
  std::sort(sorted_index.begin(), sorted_index.end(), [&required_resources](int b, int a) {
```

Reply:
Isn't required_resource already a const & ref?

Reply:
I believe when you capture it, it is copied again

Reply:
You are right. It ignores type (ptr/ref/value), and applies =/& to underlying entity.